collect_app/src/main/java/org/odk/collect/android/formmanagement/CollectFormEntryControllerFactory.kt

```diff
@@ -26,9 +36,40 @@ class CollectFormEntryControllerFactory(
 
         return FormEntryController(FormEntryModel(formDef)).also {
             val externalDataHandlerPull = ExternalDataHandlerPull(externalDataManager)
-            it.addFunctionHandler(PullDataFunctionHandler(entitiesRepository, externalDataHandlerPull))
+            it.addFunctionHandler(
+                PullDataFunctionHandler(
+                    entitiesRepository,
+                    externalDataHandlerPull
+                )
+            )
             it.addPostProcessor(EntityFormFinalizationProcessor())
+
+            if (settings.getBoolean(ProjectKeys.KEY_DEBUG_FILTERS)) {
+                it.addFilterStrategy(LoggingFilterStrategy())
+            }
+
             it.addFilterStrategy(LocalEntitiesFilterStrategy(entitiesRepository))
         }
     }
 }
+
+private class LoggingFilterStrategy : FilterStrategy {
+    override fun filter(
+        sourceInstance: DataInstance<*>,
+        nodeSet: TreeReference,
+        predicate: XPathExpression,
+        children: MutableList<TreeReference>,
+        evaluationContext: EvaluationContext,
+        next: Supplier<MutableList<TreeReference>>
+    ): MutableList<TreeReference> {
+        val startTime = System.currentTimeMillis()
+        val result = next.get()
+
+        val filterTime = System.currentTimeMillis() - startTime
```

Comment:
You can take advantage of `measureTimeMillis` function.
It could be like:
```
val result: MutableList<TreeReference>
val filterTime = measureTimeMillis { result = next.get() }
```